sky/packages/sky/lib/src/gestures/velocity_tracker.dart

```diff
@@ -0,0 +1,234 @@
+      if (xFit != null) {
+        LeastSquaresSolver ySolver = new LeastSquaresSolver(time, y, w);
+        PolynomialFit yFit = ySolver.solve(n);
+        if (yFit != null) {
+          estimator.xcoeff = xFit.coefficients;
+          estimator.ycoeff = yFit.coefficients;
+          estimator.time = newest_movement.event_time;
+          estimator.degree = n;
+          estimator.confidence = xFit.confidence * yFit.confidence;
+          return true;
+        }
+      }
+    }
+
+    // No velocity data available for this pointer, but we do have its current
+    // position.
+    estimator.xcoeff = [ x[0] ];
+    estimator.ycoeff = [ y[0] ];
+    estimator.time = newest_movement.event_time;
+    estimator.degree = 0;
+    estimator.confidence = 1.0;
+    return true;
+  }
+
+  void clear() {
+    _index = -1;
+  }
+
+  double _chooseWeight(int index) {
+    switch (weighting) {
+      case Weighting.WEIGHTING_DELTA:
+        // Weight points based on how much time elapsed between them and the next
+        // point so that points that "cover" a shorter time span are weighed less.
+        //   delta  0ms: 0.5
+        //   delta 10ms: 1.0
+        if (index == _index) {
+          return 1.0;
+        }
+        int next_index = (index + 1) % kHistorySize;
+        double delta_millis = _movements[next_index].event_time -
+          _movements[index].event_time;
+        if (delta_millis < 0)
+          return 0.5;
+        if (delta_millis < 10)
+          return 0.5 + delta_millis * 0.05;
+
+        return 1.0;
+
+      case Weighting.WEIGHTING_CENTRAL:
+        // Weight points based on their age, weighing very recent and very old
+        // points less.
+        //   age  0ms: 0.5
+        //   age 10ms: 1.0
+        //   age 50ms: 1.0
+        //   age 60ms: 0.5
+        double age_millis = _movements[_index].event_time -
+          _movements[index].event_time;
+        if (age_millis < 0)
+          return 0.5;
+        if (age_millis < 10)
+          return 0.5 + age_millis * 0.05;
+        if (age_millis < 50)
+          return 1.0;
+        if (age_millis < 60)
+          return 0.5 + (60 - age_millis) * 0.05;
+
+        return 0.5;
+
+      case Weighting.WEIGHTING_RECENT:
+        // Weight points based on their age, weighing older points less.
+        //   age   0ms: 1.0
+        //   age  50ms: 1.0
+        //   age 100ms: 0.5
+        double age_millis = _movements[_index].event_time -
+          _movements[index].event_time;
+        if (age_millis < 50) {
+          return 1.0;
+        }
+        if (age_millis < 100) {
+          return 0.5 + (100 - age_millis) * 0.01;
+        }
+        return 0.5;
+
+      case Weighting.WEIGHTING_NONE:
+      default:
+        return 1.0;
+    }
+  }
+
+  Movement _getMovement(int i) {
+      Movement result = _movements[i];
+      if (result == null) {
+        result = new Movement();
+        _movements[i] = result;
+      }
+      return result;
+  }
+
+}
+
+class VelocityTracker {
```

Comment:
I wonder how many of these classes should be private.  I'd be inclined to make them all private except this one.